File: GameStoreAndroid/app/src/main/java/johnny/tutorial/gamestoreandroid/ProductDetailActivity.java
package johnny.tutorial.gamestoreandroid;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import java.io.FileNotFoundException;

import johnny.tutorial.gamestoreandroid.constant.GameStoreConstants;
import johnny.tutorial.gamestoreandroid.constant.GameStoreConstants.ActionMode;
import johnny.tutorial.gamestoreandroid.db.ProductDbHelper;
import johnny.tutorial.gamestoreandroid.model.Product;

public class ProductDetailActivity extends AppCompatActivity {
    private static final String TAG = "ProductDetailActivity";
    private ProductDbHelper mHelper;
    private Product mProduct;
    private ActionMode mAction;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_product_detail);

        mHelper = new ProductDbHelper(this);

        Bundle extras = getIntent().getExtras();
        String actionText = extras.getString(GameStoreConstants.ParamAction);

        EditText name = (EditText) this.findViewById(R.id.productname);
        EditText price = (EditText) this.findViewById(R.id.price);
        Button loadImage = (Button)findViewById(R.id.loadimage);
        ImageView image = (ImageView) this.findViewById(R.id.image);

        if (actionText.equalsIgnoreCase(GameStoreConstants.ParamActionAdd)) {
            mAction = ActionMode.Add;
        } else if (actionText.equalsIgnoreCase(GameStoreConstants.ParamActionEdit)) {
            mAction = ActionMode.Edit;
            int id = extras.getInt(GameStoreConstants.ParamId);
            mProduct = mHelper.getProduct(id);
            if (mProduct != null) {
                name.setText(mProduct.getProductName());
                price.setText(String.valueOf(mProduct.getPrice()));
                image.setImageBitmap(mProduct.getImage());
                Drawable drawable = new BitmapDrawable(getResources(), mProduct.getImage());
                loadImage.setBackground(drawable);
                loadImage.setText("");
            }
        }

        loadImage.setOnClickListener(new Button.OnClickListener(){
            @Override
            public void onClick(View arg0) {
            Intent intent = new Intent(Intent.ACTION_PICK,
                    android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
            startActivityForResult(intent, 0);
        }});
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.detail_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // dismiss soft keyboard
        AndroidUtils.hideKeyboard(this);

        switch (item.getItemId()) {
            case R.id.action_save:
                Log.d(TAG, "Save product");
                EditText name = (EditText) this.findViewById(R.id.productname);
                EditText price = (EditText) this.findViewById(R.id.price);
                Button loadImage = (Button)findViewById(R.id.loadimage);
                ImageView image = (ImageView) this.findViewById(R.id.image);

                String textName = name.getText().toString();
                String textPrice = price.getText().toString();
                Drawable drawable = image.getDrawable();

                if(TextUtils.isEmpty(textName)){
                    Toast.makeText(this, "Product Name is Empty", Toast.LENGTH_LONG).show();
                    name.requestFocus();
                    return false;
                }
                if(TextUtils.isEmpty(textPrice)){
                    Toast.makeText(this, "Price is Empty", Toast.LENGTH_LONG).show();
                    price.requestFocus();
                    return false;
                }
                if (drawable == null) {
                    Toast.makeText(this, "Choose a image", Toast.LENGTH_LONG).show();
                    loadImage.setFocusable(true);
                    loadImage.setFocusableInTouchMode(true); ///add this line
                    loadImage.requestFocus();
                    return false;
                }

                if (mAction == ActionMode.Add) {
                    mHelper.insertProduct(textName, Double.parseDouble(textPrice), ((BitmapDrawable)drawable).getBitmap());
                } else {
                    mHelper.updateProduct(mProduct.getProductId(), textName, Double.parseDouble(textPrice), ((BitmapDrawable)drawable).getBitmap());
                }
                Intent intent = new Intent(this, ProductListActivity.class);
                this.startActivity(intent);
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == RESULT_OK){
            Uri targetUri = data.getData();
            try {
                Bitmap bitmap = BitmapFactory.decodeStream(getContentResolver().openInputStream(targetUri));
                ImageView targetImage = (ImageView)findViewById(R.id.image);
                targetImage.setImageBitmap(bitmap);
                Button loadImage = (Button)findViewById(R.id.loadimage);
                Drawable drawable = new BitmapDrawable(getResources(), bitmap);
                loadImage.setBackground(drawable);
                loadImage.setText("");
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }
        }
    }
}

File: GameStoreXamarin/GameStoreXamarin.Android/ProductDetailActivity.cs
using System;

using Android.App;
using Android.Content;
using Android.Graphics;
using Android.Graphics.Drawables;
using Android.OS;
using Android.Util;
using Android.Views;
using Android.Views.InputMethods;
using Android.Widget;
using GameStoreXamarin.Android.Helper;
using GameStoreXamarin.Core.Models;
using static GameStoreXamarin.Android.GameStoreConstants;

namespace GameStoreXamarin.Android
{
    [Activity(Label = "ProductDetailActivity")]
    public class ProductDetailActivity : Activity
    {
        private const String TAG = "ProductDetailActivity";
        private Product mProduct;
        private DetailViewMode mAction;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);

            SetContentView(Resource.Layout.ActivityProductDetail);

            Bundle extras = Intent.Extras;
            String actionText = extras.GetString(GameStoreConstants.ParamAction);

            EditText name = (EditText)FindViewById(Resource.Id.productname);
            EditText price = (EditText)FindViewById(Resource.Id.price);
            Button loadImage = (Button)FindViewById(Resource.Id.loadimage);
            ImageView image = (ImageView)FindViewById(Resource.Id.image);

            switch (actionText)
            {
                case GameStoreConstants.ParamActionAdd:
                    mAction = DetailViewMode.Add;
                    break;
                case GameStoreConstants.ParamActionEdit:
                    mAction = DetailViewMode.Edit;
                    break;
                default:
                    break;
            }

            mProduct = new Product();
            if (mAction == DetailViewMode.Edit)
            {
                int id = extras.GetInt(GameStoreConstants.ParamId);
                mProduct = DatabaseHelper.Database.GetProduct(id);
                if (mProduct != null)
                {
                    name.Text = mProduct.ProductName;
                    price.Text = Convert.ToString(mProduct.Price);
                    Bitmap bitmap = ImageHelper.BytesToBitmap(mProduct.Image);
                    image.SetImageBitmap(bitmap);
                    Drawable drawable = new BitmapDrawable(Resources, bitmap);
                    loadImage.Background = drawable;
                    loadImage.Text = "";
                }
            }

            loadImage.Click += delegate
            {
                var imageIntent = new Intent();
                imageIntent.SetType("image/*");
                imageIntent.SetAction(Intent.ActionGetContent);
                StartActivityForResult(
                    Intent.CreateChooser(imageIntent, "Select photo"), 0);
                //Intent intent = new Intent(Intent.ActionPick, MediaStore.Images.Media.ExternalContentUri);
                //StartActivityForResult(intent, 0);
            };
        }

        public override Boolean OnCreateOptionsMenu(IMenu menu)
        {
            MenuInflater.Inflate(Resource.Menu.detail_menu, menu);
            return base.OnCreateOptionsMenu(menu);
        }

        public override Boolean OnOptionsItemSelected(IMenuItem item)
        {
            EditText name1 = (EditText)FindViewById(Resource.Id.productname);
            EditText price1 = (EditText)FindViewById(Resource.Id.price);
            // dismiss soft keyboard
            InputMethodManager imm = (InputMethodManager)GetSystemService(Context.InputMethodService);
            imm.HideSoftInputFromWindow(name1.WindowToken, 0);
            imm.HideSoftInputFromWindow(price1.WindowToken, 0);

            switch (item.ItemId)
            {
                case Resource.Id.action_save:
                    Log.Debug(TAG, "Save product");
                    EditText name = (EditText)FindViewById(Resource.Id.productname);
                    EditText price = (EditText)FindViewById(Resource.Id.price);
                    Button loadImage = (Button)FindViewById(Resource.Id.loadimage);
                    ImageView image = (ImageView)FindViewById(Resource.Id.image);

                    String textName = name.Text;
                    String textPrice = price.Text;
                    Drawable drawable = image.Drawable;

                    if (String.IsNullOrEmpty(textName))
                    {
                        Toast.MakeText(this, "Product Name is Empty", ToastLength.Long).Show();
                        name.RequestFocus();
                        return false;
                    }
                    if (String.IsNullOrEmpty(textPrice))
                    {
                        Toast.MakeText(this, "Price is Empty", ToastLength.Long).Show();
                        price.RequestFocus();
                        return false;
                    }
                    if (drawable == null)
                    {
                        Toast.MakeText(this, "Choose a image", ToastLength.Long).Show();
                        loadImage.Focusable = true;
                        loadImage.RequestFocus();
                        return false;
                    }

                    mProduct.ProductName = textName;
                    mProduct.Price = Convert.ToDouble(textPrice);
                    mProduct.Image = ImageHelper.BitmapToBytes(((BitmapDrawable)drawable).Bitmap);
                    DatabaseHelper.Database.SaveProduct(mProduct);

                    var intent = new Intent(this, typeof(ProductListActivity));
                    this.StartActivity(intent);

                    return true;
                default:
                    return base.OnOptionsItemSelected(item);
            }
        }

        protected override void OnActivityResult(int requestCode, Result resultCode, Intent data)
        {
            base.OnActivityResult(requestCode, resultCode, data);

            if (resultCode == Result.Ok)
            {
                var imageView = FindViewById<ImageView>(Resource.Id.image);
                imageView.SetImageURI(data.Data);
                Button loadImage = (Button)FindViewById(Resource.Id.loadimage);
                loadImage.Background = imageView.Drawable;
                loadImage.Text = "";
                /*
                Android.Net.Uri targetUri = data.Data;
                try
                {
                    Bitmap bitmap = BitmapFactory.DecodeStream(ContentResolver.OpenInputStream(targetUri));
                    ImageView targetImage = (ImageView)FindViewById(Resource.Id.image);
                    targetImage.SetImageBitmap(bitmap);
                    Button loadImage = (Button)FindViewById(Resource.Id.loadimage);
                    Drawable drawable = new BitmapDrawable(Resources, bitmap);
                    loadImage.Background = drawable;
                    loadImage.Text = "";
                }
                catch (FileNotFoundException e)
                {
                    e.PrintStackTrace();
                }*/
            }
        }
    }
}
